go/vt/discovery/endpointwatcher.go

```diff
@@ -0,0 +1,104 @@
+package discovery
+
+import (
+	"time"
+
+	log "github.com/golang/glog"
+	pbt "github.com/youtube/vitess/go/vt/proto/topodata"
+	"github.com/youtube/vitess/go/vt/topo"
+	"golang.org/x/net/context"
+)
+
+// NewEndPointWatcher returns an EndPointWatcher and starts refreshing.
+func NewEndPointWatcher(topoServer topo.Server, hc HealthCheck, cellsToWatch []string, refreshInterval time.Duration) *EndPointWatcher {
+	epw := &EndPointWatcher{
+		topoServer:      topoServer,
+		hc:              hc,
+		cellsToWatch:    cellsToWatch,
+		refreshInterval: refreshInterval,
+		endPoints:       make(map[string]*pbt.EndPoint),
+	}
+	epw.start()
+	return epw
+}
+
+// EndPointWatcher pulls endpoints of all running tablets periodically.
+type EndPointWatcher struct {
+	topoServer      topo.Server
+	hc              HealthCheck
+	cellsToWatch    []string
+	refreshInterval time.Duration
+	cancelFunc      context.CancelFunc
+	endPoints       map[string]*pbt.EndPoint
+}
+
+// start starts the pulling.
+func (epw *EndPointWatcher) start() {
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	epw.cancelFunc = cancelFunc
+	go epw.watch(ctx)
+}
+
+// watch pulls all endpoints and notifies HealthCheck by adding/removing endpoints.
+func (epw *EndPointWatcher) watch(ctx context.Context) {
+	ticker := time.NewTicker(epw.refreshInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			watchEndPoints := make(map[string]*pbt.EndPoint)
+			for _, cell := range epw.cellsToWatch {
```

Comment:
You need to use some parallelism here. Otherwise topo server read latency is going to kill you. I'd parallelize per cell for sure. For the tablets, I'd also parallelize everything, and use a semaphore to know how many you want to read concurrently. With a flag to set the concurrency, probably 16 or 32 by default.